codex/blockexchange/engine/engine.nim

```diff
@@ -279,50 +347,54 @@ proc scheduleTasks(b: BlockExcEngine, blocks: seq[bt.Block]) {.async.} =
 
           break # do next peer
 
-proc resolveBlocks*(b: BlockExcEngine, blocks: seq[bt.Block]) {.async.} =
-  trace "Resolving blocks", blocks = blocks.len
+proc resolveBlocks*(b: BlockExcEngine, blocksDelivery: seq[BlockDelivery]) {.async.} =
+  trace "Resolving blocks", blocks = blocksDelivery.len
+
+  b.pendingBlocks.resolve(blocksDelivery)
+  await b.scheduleTasks(blocksDelivery)
+  b.discovery.queueProvideBlocksReq(blocksDelivery.mapIt( it.blk.cid ))
 
-  b.pendingBlocks.resolve(blocks)
-  await b.scheduleTasks(blocks)
-  b.discovery.queueProvideBlocksReq(blocks.mapIt( it.cid ))
+proc resolveBlocks*(b: BlockExcEngine, blocks: seq[Block]) {.async.} =
+  await b.resolveBlocks(blocks.mapIt(BlockDelivery(blk: it, address: BlockAddress(leaf: false, cid: it.cid))))
 
 proc payForBlocks(engine: BlockExcEngine,
                   peer: BlockExcPeerCtx,
-                  blocks: seq[bt.Block]) {.async.} =
-  trace "Paying for blocks", blocks = blocks.len
+                  blocksDelivery: seq[BlockDelivery]) {.async.} =
+  trace "Paying for blocks", len = blocksDelivery.len
 
   let
     sendPayment = engine.network.request.sendPayment
-    price = peer.price(blocks.mapIt(it.cid))
+    price = peer.price(blocksDelivery.mapIt(it.address))
 
   if payment =? engine.wallet.pay(peer, price):
     trace "Sending payment for blocks", price
     await sendPayment(peer.id, payment)
 
-proc blocksHandler*(
+proc blocksDeliveryHandler*(
   b: BlockExcEngine,
   peer: PeerId,
-  blocks: seq[bt.Block]) {.async.} =
-  trace "Got blocks from peer", peer, len = blocks.len
-  for blk in blocks:
-    if isErr (await b.localStore.putBlock(blk)):
-      trace "Unable to store block", cid = blk.cid
+  blocksDelivery: seq[BlockDelivery]) {.async.} =
+  trace "Got blocks from peer", peer, len = blocksDelivery.len
+
+  for bd in blocksDelivery:
+    if isErr (await b.localStore.putBlock(bd.blk)):
+      trace "Unable to store block", cid = bd.blk.cid
```

Comment:
Shouldn't we filter out failed blocks from the `blocksDelivery`?

Reply:
Hmm, that's good question. I think we can filter it out for now. It's a bit of an edge case - you can imagine a scenario where we can't put a block to the local store - maybe because it's full - but we can still process the block (i.e. stream it to a tcp socket of the user requesting the dataset via the rest api).